### DISCUSSION TITLE
How to nest multiple Span?

### DISCUSSION THREAD

**Discussion by holicc** [tokio-rs/tracing-opentelemetry#158] [Q&A] [answered]
this code can not nest `Span` instead Create a new span without relation
```
let parent_cx =
        opentelemetry::global::get_text_map_propagator(|propagator| propagator.extract(&header));
let span = tracing::info_span!(target: "otel::tracing", "Request");

span.set_parent(parent_cx.clone());
let _enter = span.enter();

    {
        let test_span = tracing::info_span!("TEST");
        test_span.set_parent(parent_cx);
        let _enter = test_span.enter();
        tracing::info!("test span");
    }
```

![image](https://github.com/user-attachments/assets/485431c3-e174-49e4-b677-6d95848e2993)

**Comment by ymgyt** [comment]
How about the following code?
```rust
 {
        let test_span = tracing::info_span!("TEST");
        // test_span.set_parent(parent_cx);
        let _enter = test_span.enter();
        tracing::info!("test span");
 }
```

**Reply by holicc** [reply]
Yes,it works. But why?  I need to set the parent context.

**Reply by ymgyt** [reply]
The parent-child relationship of spans is managed by the `tracing-subscriber`'s [`Registry`](https://github.com/tokio-rs/tracing/blob/0114ec1cf56e01e79b2e429e77c660457711d263/tracing-subscriber/src/registry/sharded.rs#L94), so even without specifying a parent, the span on the stack is maintained as the parent span. 
The information is maintained in the hierarchy `parent_cx <- span <- test_span`. Wouldn't this meet the requirements?

**Reply by holicc** [reply]
> The parent-child relationship of spans is managed by the `tracing-subscriber`'s [`Registry`](https://github.com/tokio-rs/tracing/blob/0114ec1cf56e01e79b2e429e77c660457711d263/tracing-subscriber/src/registry/sharded.rs#L94), so even without specifying a parent, the span on the stack is maintained as the parent span. The information is maintained in the hierarchy `parent_cx <- span <- test_span`. Wouldn't this meet the requirements?

Yes,  but I got two distinct spans.
- `span`
- `test_span`

You mean I shouldn't `set_parent` and instead let the `Register` manage itself?

**Comment by holicc** [comment, marked as answer]
Oh, I know why the spans are not nested. It's because I created a span first and did not set its parent, so I got `root_span <- span (with new context) <- test_span`. I broke the hierarchy of `Register`. So I created a `root_span` with context to fix the problem.
